### PR TITLE
Fixed Enum usage in PK expressions.

### PR DESCRIPTION
**Pull request by ludv1x**
Added check for duplicates of column names in CREATE query.
 [#METR-23473]

### CODE REVIEW

**Comment by ludv1x** [comment]
#164

File: dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 1]
```diff
@@ -193,9 +194,11 @@ BlockInputStreams MergeTreeDataSelectExecutor::read(
 
 	SortDescription sort_descr = data.getSortDescription();
 
-	PKCondition key_condition(query, context, available_real_and_virtual_columns, sort_descr);
+	Block pk_sample_block = data.getPrimaryExpression()->getSampleBlock();
+
+	PKCondition key_condition(query, context, available_real_and_virtual_columns, sort_descr, pk_sample_block);
 	PKCondition date_condition(query, context, available_real_and_virtual_columns,
-		SortDescription(1, SortColumnDescription(data.date_column_name, 1)));
+		SortDescription(1, SortColumnDescription(data.date_column_name, 1)), pk_sample_block);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp]: Why we are passing `pk_sample_block` to date condition?

File: dbms/src/Storages/MergeTree/PKCondition.cpp [hunk 2]
```diff
@@ -348,6 +353,47 @@ bool PKCondition::isPrimaryKeyPossiblyWrappedByMonotonicFunctionsImpl(
...
+	/// Try to correct type of constant for correct comparison
+	try
+	{
+		/// Convert String to Enum's value
+		if (auto data_type_enum = dynamic_cast<const IDataTypeEnum *>(desired_type.get()))
+		{
+			src_value = data_type_enum->castToValue(src_value);
+		}
+		/// Convert 'YYYY-MM-DD' Strings to Date
+		else if (typeid_cast<const DataTypeDate *>(desired_type.get()))
```
Comment (alexey-milovidov) [inline dbms/src/Storages/MergeTree/PKCondition.cpp]: What about DateTime?

File: dbms/src/Storages/MergeTree/PKCondition.cpp [hunk 3]
```diff
@@ -348,6 +353,47 @@ bool PKCondition::isPrimaryKeyPossiblyWrappedByMonotonicFunctionsImpl(
...
+			/// Ok, numeric types are mutually convertible
+		}
+		else
+		{
+			return false;
+		}
+	}
+	catch (...)
+	{
+		return false;
```
Comment (alexey-milovidov) [inline dbms/src/Storages/MergeTree/PKCondition.cpp]: There was old behaviour where exception is thrown immediately for comparison of Date/DateTime with string literal that contains invalid date. Is that behaviour has changed?

**Comment by ludv1x** [comment]
Updated.

**Comment by ludv1x** [comment]
BTW Also, there are need to add accurate comparisons for non-64-bit types to make filtering py PK consistent with filtering through full scan.

**Pull request merged**